### DISCUSSION TITLE
Time stepper for very stiff systems

### DISCUSSION THREAD

**Discussion by heinono1** [idaholab/moose#21879] [Q&A General] [answered]
John Mangeri and I have been working on a micromagnetic module ('ferret') for MOOSE. Because the norm of the magnetization vector field is conserved at each point in space, we were trying to implement spherical coordinates, as this will automatically eliminate a problem of renormalizing the vector field at each time step (and reduces the DOFs). It turns out, however, that the governing equations, the Landau-Lifshitz-Gilbert equations, are very stiff, in particular in spherical coordinates. We have scoured through different time-integrators and time stepper in MOOSE, but the result always ends up being the same: the solutions become unstable after some time and start to diverge. A particularly sensitive test of solving the equations of motion is to turn the damping, 'alpha', to zero, in which case total energy should be strictly conserved. It is not in spherical coordinates - even with tiny time steps, the energy starts to diverge (increase) after some number of timesteps. It appears that the main cause of these problems is the magnetic exchange coupling. In simple test systems without exchange ('macrospins'), we can make the energy strictly conserved in the absence of damping, but the timesteps have to be tiny, otherwise the energy oscillates (but does not diverge). 

We are interested in reaching out to the MOOSE community to hear if others have had similar or analogous problems, and what they did to address it. One potential path is to use more timestepper options that are available in PETSc, or other more complicated implicit steppers, but such pathways are beyond my comfort zone to implement in MOOSE.

I suspect part of why this is a problem may have to do with the nature of weak formulations. In strong formulations I can make (in my own codes as well as others') energy conservation be respected. I am guessing that the elemental form of the nonlinear variables in weak formulation causes the problem: the delicate and precise solutions that you need to preserve norm and energy are possible in strong formulations but get washed out in weak formulation.

I would be really interested in hearing comments and suggestions from the MOOSE community!

**Comment by GiudGiud** [comment]
@cticenhour

**Reply by GiudGiud** [reply]
@idaholab/moose-team 
@hugary1995 @sapitts

**Comment by roystgnr** [comment]
Pardon my ignorant questions; I don't think I can help but there's a slim chance I could elicit information that might be enough for someone else to help.

Can you post the weak formulation you're using and the FE space you've chosen?  Also, is there a particular mode you're seeing in the diverging solutions?

Also, by spherical coordinates, do you mean that instead of solving for the evolution of M_x,M_y,M_z, you're trying to solve for M_theta and M_phi?  If so ... is there even an upper bound on the stiffness you get from that?  As the polar angle goes to zero, the sensitivity of the azimuthal angle to the actual vector position goes to infinity.  The problem is like gimbal lock, but the solutions I've heard of for gimbal lock (rotation matrices, quaternions) involve exactly the sort of "extra DoFs + renormalization" that you're trying to avoid!

**Comment by heinono1** [comment]
Hello,

You are correct that we are trying to solve for m_theta and m_phi instead
of for the Cartesian components. What I have been using for test problems
very carefully avoid getting theta close to either pole in order to avoid
issues with approaching singularities at the poles.

I am not sure that there is a particular mode that diverges - I have not
done a careful instability analysis. Heuristically, in these small test
systems, a few spins start to get unstable and just take off to stupid
values.

The weak formulation is just a standard weak formulation of the
Landau-Lifshitz-Gilbert equations, with appropriate partial integrations on
the contribution to residuals from the exchange field to move one
derivative order from the magnetization vector field to the test functions.
I can send you more details if you want.

Cheers,
Olle

On Thu, Aug 18, 2022 at 4:41 PM roystgnr ***@***.***> wrote:

> Pardon my ignorant questions; I don't think I can help but there's a slim
> chance I could elicit information that might be enough for someone else to
> help.
>
> Can you post the weak formulation you're using and the FE space you've
> chosen? Also, is there a particular mode you're seeing in the diverging
> solutions?
>
> Also, by spherical coordinates, do you mean that instead of solving for
> the evolution of M_x,M_y,M_z, you're trying to solve for M_theta and M_phi?
> If so ... is there even an upper bound on the stiffness you get from that?
> As the polar angle goes to zero, the sensitivity of the azimuthal angle to
> the actual vector position goes to infinity. The problem is like gimbal
> lock, but the solutions I've heard of for gimbal lock (rotation matrices,
> quaternions) involve exactly the sort of "extra DoFs + renormalization"
> that you're trying to avoid!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/21879#discussioncomment-3426896>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF7TWE3WXOKEXPCTZXDVZ2UZRANCNFSM565RIRGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Reply by lindsayad** [reply]
What time integration scheme is typically used in the literature for these problems?

**Comment by heinono1** [comment]
Good question. In Cartesian coordinates usually higher-order (explicit)
Runge-Kutta. I use an implicit Bulirch-Stoer method. For spherical
coordinates, the literature is pretty quiet as far as I know. There is an
old code, MagPar (you can find it on gitHub), that uses spherical
coordinates and BDF stepper from SUNDIAL but apparently that does not
work so well for a lot of problems.

On Fri, Aug 19, 2022 at 3:23 AM Alex Lindsay ***@***.***>
wrote:

> What time integration scheme is typically used in the literature for these
> problems?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/21879#discussioncomment-3429452>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF22ZZMEFICPHIKDBUDVZ5AABANCNFSM565RIRGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Reply by lindsayad** [reply]
Have you tried BDF2 in MOOSE? It sounds like maybe you have? When the energy starts to diverge, is that with a constant time step?

**Reply by roystgnr** [reply]
That's surprising.  I know a change of variables can screw up your discrete stability, but if your solution is staying well away from the pole singularities I don't see how the stability could easily go from "fits in the small bounded stability region promised by RK" to "unstable even with an implicit method".  I'm not familiar with Bulirch-Stoer ... but it's in the explicit volume of Hairer & Wanner, and the examples I can find online look explicit too.  You're using a version that's built around the implicit midpoint method instead of the explicit midpoint method?  Any idea what the stability domain looks like?

**Comment by heinono1** [comment]
Hi Guillaume,

Yes, I have tried BDF2 with adaptive steps. But I have also set the dtmax
to some ridiculously small time step - system still goes nuts. The problems
start not with difficulties diverging but with some number of spins
becoming unstable (exchange coupling should keep the spin texture smooth
but some spins start to deviate from the smoothness).

-Olle

On Tue, Aug 23, 2022 at 5:59 AM Alex Lindsay ***@***.***>
wrote:

> Have you tried BDF2 in MOOSE? It sounds like maybe you have? When the
> energy starts to diverge, is that with a constant time step?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/21879#discussioncomment-3455526>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF4GSTMGLBWBG2YCY7LV2SVIXANCNFSM565RIRGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Reply by GiudGiud** [reply]
linsdayad is actually Alex, I'm Guillaume

Is there some literature about stabilizing those solves?
Could you add some sort of diffusive term to those equations?

We also have Runge Kunta schemes at various orders btw. 

When you say some spins become unstable, how do you diagnose that? Are they a separate variable?

**Comment by heinono1** [comment]
Hi,

The spins are the nonlinear variables. We monitor the instability in
different ways. Typically, the exchange starts to grow (some neighboring
spins start to become misaligned) and visually we confirm that misalignment
of neighboring spins starts. I have tried a variety of steppers. Typically,
the Newmarkbeta works the best for when we use a Cartesian representation
but that one too becomes unstable in spherical representation.

I am not aware of any particular literature on how to stabilize the solvers
for spherical coordinates. It is perhaps possible to add a diffusive term
to stabilize the equations but it has to be done such that the equations of
motion (the physics) are not made unphysical.



On Tue, Aug 23, 2022 at 9:09 AM Guillaume Giudicelli <
***@***.***> wrote:

> linsdayad is actually Alex, I'm Guillaume
>
> Is there some literature about stabilizing those solves?
> Could you add some sort of diffusive term to those equations?
>
> We also have Runge Kunta schemes at various orders btw.
>
> When you say some spins become unstable, how do you diagnose that? Are
> they a separate variable?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/21879#discussioncomment-3457092>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF3I6QQSTGTZ7DLQWJDV2TLP5ANCNFSM565RIRGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Reply by GiudGiud** [reply]
how big are the numerical systems btw?
And the linear and nonlinear solves are fine, you never just accept an insufficiently converged solution, this is only a time stepping issue?

**Comment by heinono1** [comment]
The numerical systems range from some 10k DOFs to many (many) millions of
DOFs. Yes, the linear and nonlinear solvers are fine, it seems to be
primarily a time-stepping issue. Of course, as some spin DOFs start to get
wonky it becomes difficult for the linear and nonlinear solvers to
converge, but I believe that is a secondary problem.

On Tue, Aug 23, 2022 at 11:14 AM Guillaume Giudicelli <
***@***.***> wrote:

> how big are the numerical systems btw?
> And the linear and nonlinear solves are fine, you never just accept an
> insufficiently converged solution, this is only a time stepping issue?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/21879#discussioncomment-3458266>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF7MSJ75CW7WJOZ4TBLV2T2HLANCNFSM565RIRGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Reply by GiudGiud** [reply]
at 10k DOFs, with some patience, we may be able to use the svd_monitor to check that the problems are actually well conditioned

If this is strictly a time-stepping issue, I think we should look at a fourier analysis of the system, then look at which schemes should be stable and why they are not.

**Reply by hugary1995** [reply]
I am not familiar with this particular physics you are looking at. But like you said, it sounds like the literature is scarce in this direction. So treating this as a research question, I would first sit down and go through the stability analysis for this discretization. It may very well be the case that the inf-sup condition is okay in cartesian but not in other coordinates.

**Comment by heinono1** [comment]
Hi Gary and Roy,

Thanks for your comments. The Bulirch-Stoer stepper I built (for
finite-difference and FFT methods) is an implicit one, pretty much straight
out of Numerical Recipes - for Cartesian coordinates. I have not tried it
for spherical coordinates.

I have no idea about stability regions - I have not done any stability
analyses.

Cheers,
Olle

On Tue, Aug 23, 2022 at 3:45 PM roystgnr ***@***.***> wrote:

> That's surprising. I know a change of variables can screw up your discrete
> stability, but if your solution is staying well away from the pole
> singularities I don't see how the stability could easily go from "fits in
> the small bounded stability region promised by RK" to "unstable even with
> an implicit method". I'm not familiar with Bulirch-Stoer ... but it's in
> the explicit volume of Hairer & Wanner, and the examples I can find online
> look explicit too. You're using a version that's built around the implicit
> midpoint method instead of the explicit midpoint method? Any idea what the
> stability domain looks like?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/21879#discussioncomment-3460162>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF3DRCKFGEDE3ISQ4A3V2UZ7LANCNFSM565RIRGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Comment by mangerij** [comment]
I like the idea of using svd-monitor.

I will try this for a small problem with Cartesian (working just doesn't scale the dof that Olle and company need) and also the spherical formulation (doesn't work/converge at any scale).

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by GiudGiud** [reply]
please let us know how it goes. 
I've seen a lot of time-stepping give insufficiently accurate answers but not with a refined timestep, hence the idea to have a look at other parts of the solve

**Reply by GiudGiud** [reply]
Any update on this?

**Comment by heinono1** [comment]
Hi Guillaume,

No, have not had time to look into this more - have worked on the
Cartiesian version (which seems to be chugging along pretty nicely now).

CHeers,
Olle

On Thu, Sep 29, 2022 at 11:29 AM Guillaume Giudicelli <
***@***.***> wrote:

> Any update on this?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/21879#discussioncomment-3764457>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEFYLXQLRWQKI63QXUQTWAW7X3ANCNFSM565RIRGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Comment by mangerij** [comment, marked as answer]
also, I will provide a .pdf for the governing equations, residuals, and jacobians for the simple problem in a few days (maybe next week)

Cheers
J

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by GiudGiud** [reply]
Should we close this?
Where are we on this particular issue

**Reply by heinono1** [reply]
Ok to close as far as I am concerned. Would be nice to have a solution developed but it would take a lot of work, I think, and I don't have bandwidth to deal with this. The Cartesian+renormalization works for me.
